File: frontend/src/components/benchmark/content/betweenness_centrality/implementations/betweenness_centrality.java
import java.util.*;

public class BetweennessCentrality {
    private Map<Integer, Set<Integer>> adjacencyList;

    public BetweennessCentrality() {
        this.adjacencyList = new HashMap<>();
    }

    public void addEdge(int u, int v) {
        adjacencyList.putIfAbsent(u, new HashSet<>());
        adjacencyList.putIfAbsent(v, new HashSet<>());
        adjacencyList.get(u).add(v);
        adjacencyList.get(v).add(u);
    }

    public void buildFromEdges(List<int[]> edges) {
        for (int[] edge : edges) {
            addEdge(edge[0], edge[1]);
        }
    }

    public Map<Integer, Double> betweennessCentrality() {
        Map<Integer, Double> centrality = new HashMap<>();
        for (int node : adjacencyList.keySet()) {
            centrality.put(node, 0.0);
        }

        for (int s : adjacencyList.keySet()) {
            Stack<Integer> stack = new Stack<>();
            Map<Integer, List<Integer>> pred = new HashMap<>();
            Map<Integer, Integer> dist = new HashMap<>();
            Map<Integer, Integer> sigma = new HashMap<>();

            for (int node : adjacencyList.keySet()) {
                pred.put(node, new ArrayList<>());
                dist.put(node, -1);
                sigma.put(node, 0);
            }

            dist.put(s, 0);
            sigma.put(s, 1);
            Queue<Integer> queue = new LinkedList<>();
            queue.add(s);

            while (!queue.isEmpty()) {
                int v = queue.poll();
                stack.push(v);

                for (int w : adjacencyList.get(v)) {
                    if (dist.get(w) < 0) {
                        dist.put(w, dist.get(v) + 1);
                        queue.add(w);
                    }
                    if (dist.get(w) == dist.get(v) + 1) {
                        sigma.put(w, sigma.get(w) + sigma.get(v));
                        pred.get(w).add(v);
                    }
                }
            }

            Map<Integer, Double> delta = new HashMap<>();
            for (int node : adjacencyList.keySet()) {
                delta.put(node, 0.0);
            }

            while (!stack.isEmpty()) {
                int w = stack.pop();

                for (int v : pred.get(w)) {
                    delta.put(v, delta.get(v) + ((double)sigma.get(v) / sigma.get(w)) * (1 + delta.get(w)));
                }

                if (w != s) {
                    centrality.put(w, centrality.get(w) + delta.get(w));
                }
            }
        }

        return centrality;
    }

    public List<Map.Entry<Integer, Double>> topKCentralNodes(int k) {
        Map<Integer, Double> centrality = betweennessCentrality();
        List<Map.Entry<Integer, Double>> result = new ArrayList<>(centrality.entrySet());
        
        result.sort((a, b) -> b.getValue().compareTo(a.getValue()));
        
        return result.subList(0, Math.min(k, result.size()));
    }
}

File: frontend/src/components/benchmark/content/betweenness_centrality/implementations/betweenness_centrality.js
class BetweennessCentrality {
    constructor() {
        this.adjacencyList = new Map();
    }

    addEdge(u, v) {
        if (!this.adjacencyList.has(u)) {
            this.adjacencyList.set(u, new Set());
        }
        if (!this.adjacencyList.has(v)) {
            this.adjacencyList.set(v, new Set());
        }
        this.adjacencyList.get(u).add(v);
        this.adjacencyList.get(v).add(u);
    }

    buildFromEdges(edges) {
        for (const [u, v] of edges) {
            this.addEdge(u, v);
        }
    }

    betweennessCentrality() {
        const centrality = {};
        for (const node of this.adjacencyList.keys()) {
            centrality[node] = 0.0;
        }

        for (const s of this.adjacencyList.keys()) {
            const stack = [];
            const pred = {};
            const dist = {};
            const sigma = {};

            for (const node of this.adjacencyList.keys()) {
                pred[node] = [];
                dist[node] = -1;
                sigma[node] = 0;
            }

            dist[s] = 0;
            sigma[s] = 1;
            const queue = [s];

            while (queue.length > 0) {
                const v = queue.shift();
                stack.push(v);

                for (const w of this.adjacencyList.get(v)) {
                    if (dist[w] < 0) {
                        dist[w] = dist[v] + 1;
                        queue.push(w);
                    }
                    if (dist[w] === dist[v] + 1) {
                        sigma[w] += sigma[v];
                        pred[w].push(v);
                    }
                }
            }

            const delta = {};
            for (const node of this.adjacencyList.keys()) {
                delta[node] = 0.0;
            }

            while (stack.length > 0) {
                const w = stack.pop();

                for (const v of pred[w]) {
                    delta[v] += (sigma[v] / sigma[w]) * (1 + delta[w]);
                }

                if (w !== s) {
                    centrality[w] += delta[w];
                }
            }
        }

        return centrality;
    }

    topKCentralNodes(k = 10) {
        const centrality = this.betweennessCentrality();
        const result = Object.entries(centrality)
            .sort((a, b) => b[1] - a[1])
            .slice(0, k)
            .map(([node, score]) => ({ node: parseInt(node), score }));
        
        return result;
    }
}

module.exports = BetweennessCentrality;